Wizard component

dlabrecq opened the pull request:
Adding the wizard component, based on the Angular Patternfly implementation with some refactoring, bug fixes, etc. For example, patternfly-ng favors emitting events (for functionality such as next, previous, close, etc.) instead of relying on user provided callbacks.

This PR includes:

- Components for wizard, wizard-step, wizard-substep, and wizard-review.
- Classes for wizard-config, wizard-step-config, and wizard-event.
- A wizard-base class for common wizard and wizard-step functionality.
- A wizard-step interface for step and substep types.
- A modal example
- A modal example with paused navigation (a Cloud Forms requirement)
- An embed in page example showing a slightly different button layout

Note: Angular Patternfly does not provide unit tests for wizard. Need to figure out how to test ng-templates properly, but not for this go around -- added to issue https://github.com/patternfly/patternfly-ng/issues/31.

Live Example:
https://rawgit.com/dlabrecq/patternfly-ng/wizard-dist/dist-demo/#/wizard

Modal Example:
![screen shot 2017-08-04 at 11 49 12 am](https://user-images.githubusercontent.com/17481322/28976204-043550b8-790b-11e7-8350-3b3b27f93079.png)

Embed in Page Example:
![screen shot 2017-08-04 at 11 50 27 am](https://user-images.githubusercontent.com/17481322/28976252-2c8bc88a-790b-11e7-9d53-c3426a8a952d.png)

dlabrecq committed:
feat(wizard): Added wizard, including:

- Components for wizard, wizard-step, wizard-substep, and wizard-review.
- Classes for wizard-config, wizard-step-config, and wizard-event.
- A wizard-base class for common wizard and wizard-step functionality.
- A wizard-step interface for step and substep types.
- A wizard example

dlabrecq commented:
FYI @sjcox-rh @catrobson

AdamJ commented on package.json:
```diff
@@ -67,8 +67,8 @@
-    "ngx-bootstrap": "1.7.1",
-    "patternfly": "^3.24.0",
+    "ngx-bootstrap": "1.8.0",
+    "patternfly": "^3.5.0",
```
should this be version `^3.25.0`?

dlabrecq replied:
Yes, thanks for finding that! Bumped to v3.26.1

dlabrecq committed:
fix(package): Bumped patternfly version number

catrobson commented:
@dlabrecq can we add labels to the steps along the top?  "First Step, Second Step, etc".  It looks like there is space for them but nothing is showing.
![image](https://user-images.githubusercontent.com/3739227/28893209-974e7974-779e-11e7-8ab0-4780660f4a83.png)

@mindreeper2420 it looks like the final step of the wizard is using the fontawesome fa-check-circle rather than the patternfly pficon-ok as specified by the pattern.  Is this inherited and needs to be fixed upstream?

I also noticed that in the implementation the final screen shows the cancel and back buttons as disabled, but the pattern shows the cancel button disappearing and the back button enabled.  I think how it is implemented is actually more accurate to normal usage.  Do you all agree?  If so I'll create an issue against PatternFly core.

dlabrecq commented:
@catrobson I've fixed the labels.

The check circle is actually using the icon below. This is the same icon used in the Patternfly test pages and for Angular Patternfly. It's actually defined by the app developer, not the wizard itself.

&lt;div class="wizard-pf-success-icon">
  &lt;span class="glyphicon glyphicon-ok-circle">&lt;/span>
&lt;/div>

Regarding the cancel button behavior, the example also mimics the Patternfly test pages and Angular Patternfly. However, it's up to the app developer to define that behavior. Thus, I can change the example to do whatever you prefer?

dlabrecq committed:
fix(wizard): Step indicator labels are not shown

dlabrecq commented:
@catrobson @mindreeper2420 What would you like to do about the ok icon and cancel button?

dlabrecq committed:
fix(wizard): Refactored to accommodate a CloudForms use case

After speaking with J. Phillips, Cloud Forms has a use case where the user has clicked next, but is blocked from navigating to the next step/substep until a process has completed. I’ve modified the example and wizard component as follows.

The step/substep allowNavAway config property is initially set to false, which keeps the user on the current step. When the user clicks the wizard’s next button, an event is emitted which an app developer can use to kick off a process. The app developer would then show a progress spinner in the step/substep main panel while the user is waiting. When the process has completed, allowNavAway is set to true and the wizard’s goToNextStep() method is called, allowing navigation to the next step/substep.

catrobson commented:
@dlabrecq I'm good to merge as is.  OK icon should be fixed as we get new versions of PF-icons in the CSS, and Cancel button works as expected for an example IMO.

dlabrecq committed:
chore(wizard): Added embed in page example

dlabrecq committed:
chore(wizard): Cleaned up examples
